gateway/src/lib.rs

```diff
@@ -683,43 +676,150 @@ impl Actor {
-                        st.commit_topdown_msg(rt.store(), &mut cross_msg)
-                            .map_err(|e| {
-                                e.downcast_default(
-                                    ExitCode::USR_ILLEGAL_STATE,
-                                    "error committing top-down message while applying it",
-                                )
-                            })?;
-                    }
-                    Ok(())
-                })?;
+                // TODO: this check should still be enforced in propagate?
+                // TODO: We might run into nonce not sequential issue. Currently we
+                // TODO: are sharing the nonce across the network, this might be slow
+                // TODO: as owners needs to execute one after anotehr and each owner
+                // TODO: might have a different speed and it might cause congestion?
+                if st.applied_topdown_nonce != cross_msg.msg.nonce {
+                    return Err(actor_error!(
+                        illegal_state,
+                        "the top-down message being applied doesn't hold the subsequent nonce"
+                    ));
+                }
 
-                // if directed to the current network propagate the message
                 if sto == st.network_name {
-                    // FIXME: Should we handle return in some way?
-                    let _ = cross_msg.send(rt, rto)?;
+                    rt.transaction(|st: &mut State, _| {
+                        st.applied_topdown_nonce += 1;
+                        Ok(())
+                    })?;
+
+                    // We can return the send result
+                    return cross_msg.send(rt, rto);
                 }
             }
             _ => {
                 return Err(actor_error!(
                     illegal_argument,
                     "cross-message to apply dosen't have the right type"
-                ));
+                ))
             }
         };
 
-        Ok(())
+        rt.transaction(|st: &mut State, rt| {
+            let owner = cross_msg
+                .msg
+                .from
+                .raw_addr()
+                .map_err(|_| actor_error!(illegal_argument, "invalid address"))?;
+            st.insert_postbox(
+                rt.store(),
+                Some(vec![owner]),
+                MIN_CROSS_MSG_GAS.clone(),
+                cross_msg,
+            )
+            .map_err(|e| {
+                e.downcast_default(ExitCode::USR_ILLEGAL_STATE, "error save topdown messages")
+            })?;
+            Ok(())
+        })?;
+
+        Ok(RawBytes::default())
+    }
+
+    fn propagate<BS, RT>(rt: &mut RT, params: PropagateParams) -> Result<RawBytes, ActorError>
+    where
+        BS: Blockstore,
+        RT: Runtime<BS>,
+    {
+        // TODO: update this to EOA
+        rt.validate_immediate_caller_accept_any()?;
+
+        let PropagateParams { gas, postbox_cid } = params;
```

Comment:
We can remove `gas` from `PropagateParams`. As discussed sync, this will be a constant set by the actor. What we should check here is if the `rt.message().amount` is over the gas set for cross-messages. This will increase the internal balance of the actor.

We should also keep track of the amount of fees collected in a variable, as this will be used in the future to distribute rewards.